### ISSUE TITLE
Use a file lock for restricting running multiple instances of the application

### ISSUE THREAD

**Issue by ksaric** [input-output-hk/cardano-shell#22]
We need to add an application-level OS lock, to detect, prevent and report that another instance of the application is still running.

We should check with Devops and the frontend team to make sure they're happy with the behavior.

Use the `GHC.IO.Handle.Lock` API. It needs an application lock file, but the lock is not just that the file exists, it's a proper OS level API that automatically unlocks the file if the process terminates. Use `hTryLock` to try to acquire the lock, so you can fail if you don't get it (whereas `hLock` blocks until it gets it).
This is based on the problem that we observe with the existing code that we suspect many of the upgrade problems are due to the old version still running. The point here would be to make that diagnosis clear and reliable, to help reduce user confusion.

For example, somebody has two installations, one in `/home/user/cardano-sl-2.0` and the other in `/home/user/cardano-sl-1.33`. Each instance of the program that runs knows which dir is its app state dir, and it uses the lock file in that dir. So, in this case, there will be two lock files, and you can run both versions concurrently, but not two instances of the same version. I guess it's better to think about it as the lock file protecting the application state, rather than about preventing multiple instances of the application from running.

Another example, somebody has two installations, one in `/home/user/cardano-sl-2.0-installation-1` and on in `/home/user/cardano-sl-2.0-installation-2`. Perhaps ports will still conflict. But it catches the primary issue with upgrades, where we don't change the state dir.

This fits in as one of your modular features in the framework, that you provide as optional bundled features. It does a check on initialization (and can fail synchronously), it can find the app state dir by config, or from the server environment. It can release the lock file on shutdown (ok, that's actually automatic, but it fits into your framework as a nice example).

**Comment by ksaric** [comment]
@iohk-devops and @nikolaglumac - can you comment on this? Is this ok with you guys?

**Comment by nikolaglumac** [comment]
@ksaric what happens if you cut the power to your machine while Cardano is still running and the lock-file is left on the disk?

**Comment by ksaric** [comment]
@nikolaglumac Good question. It should lock just the file contents, maybe the name of the issue is not that good. Will post an example to confirm.

**Comment by ksaric** [comment]
Seems to work ok, whether the process shuts down unexpectedly or not.
To test out, we first create a test file. We can call it `lockfile` (creative, I know) - `touch lockfile`.

Start two `stack ghci` processes in separate terminals.
In the first one we write:
```
λ> import GHC.IO.Handle.Lock
λ> testHandle <- openFile "lockfile" ReadMode
λ> result <- hTryLock testHandle ExclusiveLock 
λ> result
True
```
So we have an `ExclusiveLock`. Let's try it from the second terminal:
```
λ> import GHC.IO.Handle.Lock
λ> testHandle <- openFile "lockfile" ReadMode
λ> result <- hTryLock testHandle ExclusiveLock 
λ> result
False
```
:sob: 
That's what we wanted. So we shut down the terminal (anyway you like). So we try to get the resource in the second terminal:
```
λ> result' <- hTryLock testHandle ExclusiveLock
λ> result'
True
```
Yaay! It works. 
So it doesn't create a file, but uses one to synchronize access for processes. Which is exactly what we want.

**Comment by ksaric** [comment]
@cleverca22 If I got this right - as you mentioned, it should be possible to implement the installation in such a way that we make sure that there are no overwrites until reboot on Windows, right?

https://nsis.sourceforge.io/Replace_File_On_Reboot

**Comment by cleverca22** [comment]
yeah, there is the option of using that for every file being installed on windows
then you just need to keep in mind, that its forcing the user to reboot after any update or installation, which can be very anoying

**Comment by ksaric** [comment]
Awesome, I'll add an issue for this!
